### PR TITLE
[chatgpt] add supervised learning fine-tune code

### LINKED ISSUE #3182: [PROPOSAL]: [chatgpt] add code for  fine-tuning supervised model
### Proposal

Based on the instructions by InstructGPT, training the model involves three steps. ColossalAI has already implemented steps 2 and 3. Although there is code available for pretraining large language models (such as GPT) in this project, it does not fit well with ChatGPT. Therefore, I have written the code for step 1, which involves fine-tuning the model using supervised learning.

### Self-service

- [X] I'd be willing to do some initial work on this proposal myself.

### PR DESCRIPTION
**Pull request by pgzhang**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number
resolved #3182 

## 📝 What does this PR do?
Add supervised learning fine-tune code as mentioned in InstructGPT(step 1)

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/ChatGPT/chatgpt/trainer/sft.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+class SFTTrainer(ABC):
+    """
+        Trainer to use while training reward model.
+
+    Args:
+        model (torch.nn.Module): the model to train
+        strategy (Strategy): the strategy to use for training
+        optim(Optimizer): the optimizer to use for training
+        train_dataset (RewardDataset): the dataset to use for training
```
Comment (ht-zhou) [inline applications/ChatGPT/chatgpt/trainer/sft.py:24]: SFT Dataset？

File: applications/ChatGPT/examples/train_sft.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    max_len = 512
+
+    # configure optimizer
+    if args.strategy.startswith('colossalai'):
+        optim = HybridAdam(model.parameters(), lr=5e-5)
+    else:
+        optim = Adam(model.parameters(), lr=5e-5)
+
+    logger = get_dist_logger()
+    #data = load_dataset(args.dataset, 'super_natural_instructions')
```
Comment (ht-zhou) [inline applications/ChatGPT/examples/train_sft.py:68]: please delete this kind of unused code

File: applications/ChatGPT/chatgpt/trainer/sft.py (hunk 1 above)
Comment (pgzhang) [reply applications/ChatGPT/chatgpt/trainer/sft.py:24]: Yes, I borrowed some code from the origin project, and I forgot to check the comment code...

File: applications/ChatGPT/examples/train_sft.py (hunk 2 above)
Comment (pgzhang) [reply applications/ChatGPT/examples/train_sft.py:68]: ok

File: applications/ChatGPT/chatgpt/dataset/sft_dataset.py [hunk 3]
```diff
@@ -44,6 +44,8 @@ def __getitem__(self, idx):
 
 
 class DistributedSampler:
+    # This code is borrowd from:
```
Comment (ht-zhou) [inline applications/ChatGPT/chatgpt/dataset/sft_dataset.py:47]: https://pytorch.org/docs/stable/_modules/torch/utils/data/distributed.html#DistributedSampler
Can we use torch distributed sampler here?
Or if you want to use Wenet's code, we have to add license into our license and declare the copyright.

Comment (pgzhang) [reply applications/ChatGPT/chatgpt/dataset/sft_dataset.py]: I have rewritten the sampler with torch distributed sampler

**Pull request merged**